radixdlt/src/main/java/com/radixdlt/statecomputer/RadixEngineStateComputer.java

```diff
@@ -195,19 +198,47 @@ private void commitCommand(long version, Command command, VerifiedLedgerHeaderAn
 			// TODO: Include permission level in committed command
 			this.radixEngine.checkAndStore(committedAtom, PermissionLevel.SUPER_USER);
 		} catch (RadixEngineException | DeserializeException e) {
-			// TODO: Remove throwing of exception
-			// TODO: Exception could be because of byzantine quorum
-			throw new IllegalStateException("Trying to commit bad command", e);
+			throw new ByzantineQuorumException("Trying to commit bad command", e);
 		}
 	}
 
 	@Override
 	public void commit(VerifiedCommandsAndProof verifiedCommandsAndProof) {
+		final SystemParticle lastSystemParticle = radixEngine.getComputedState(SystemParticle.class);
+		final long currentEpoch = lastSystemParticle.getEpoch();
+		boolean epochChange = false;
+
 		final VerifiedLedgerHeaderAndProof headerAndProof = verifiedCommandsAndProof.getHeader();
 		long stateVersion = headerAndProof.getAccumulatorState().getStateVersion();
 		long firstVersion = stateVersion - verifiedCommandsAndProof.getCommands().size() + 1;
 		for (int i = 0; i < verifiedCommandsAndProof.getCommands().size(); i++) {
 			this.commitCommand(firstVersion + i, verifiedCommandsAndProof.getCommands().get(i), headerAndProof);
+
+			long nextEpoch = radixEngine.getComputedState(SystemParticle.class).getEpoch();
+			if (nextEpoch > currentEpoch) {
+				// TODO: revert commands and log as proof of byzantine quorum rather than throw illegal state exception
+				if (i != verifiedCommandsAndProof.getCommands().size() - 1) {
+					throw new ByzantineQuorumException("change of epoch did not occur on last command");
+				}
+				epochChange = true;
+				break;
```

Comment:
is this break needed, since epoch change can only happen in the last command anyway?

btw, I'm a big fan of exhaustive if-else statements (and trying to avoid nesting, if possible), what do you think about this version?
```
if (isLastCommand && commandChangedEpoch) {
  epochChange = true;
} else if (commandChangedEpoch) {
  throw new ByzantineQuorumException();
} else {
  continue;
}
```

Reply:
I like it, thanks